File: tasks/translation/output/Qwen/QwQ-32B-AWQ/codenet/direct/comment/uncomment/direct/Java/Python/s685734377_commented.py
import sys

def main():
    n = int(sys.stdin.readline())
    a = []
    for _ in range(n):
        line = sys.stdin.readline().strip()
        a.append(list(line))
    
    P = 37
    Q = 47
    powP = [1] * (n + 1)
    for i in range(1, n+1):
        powP[i] = powP[i-1] * P
    
    hashRow = [[0]*n for _ in range(n)]
    hashCol = [[0]*n for _ in range(n)]
    
    for i in range(n):
        for j in range(n):
            c = ord(a[i][j])
            for k in range(n):
                term = c * powP[(j - k) % n]
                hashRow[i][k] += term
                term_col = c * powP[(i - k) % n]
                hashCol[j][k] += term_col
    
    ans = 0
    for A in range(n):
        for B in range(n):
            ok = True
            for i in range(n):
                row_iA = (i + A) % n
                val_row = hashRow[row_iA][B]
                col_iB = (i + B) % n
                val_col = hashCol[col_iB][A]
                if val_row != val_col:
                    ok = False
                    break
            if ok:
                ans += 1
    print(ans)

if __name__ == "__main__":
    main()
# End of Code

File: tasks/translation/output/Qwen/QwQ-32B-AWQ/codenet/direct/javadoc/uncomment/direct/Java/C/s685734377_commented.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

int main() {
    int n;
    scanf("%d", &n);

    char **a = malloc(n * sizeof(char *));
    for (int i = 0; i < n; i++) {
        a[i] = malloc(n * sizeof(char));
    }

    char line[1024];
    for (int i = 0; i < n; i++) {
        fgets(line, sizeof(line), stdin);
        line[strcspn(line, "\n")] = '\0';
        for (int j = 0; j < n; j++) {
            a[i][j] = line[j];
        }
    }

    const long long P = 37;
    long long *powP = malloc((n + 1) * sizeof(long long));
    powP[0] = 1;
    for (int i = 1; i <= n; i++) {
        powP[i] = powP[i - 1] * P;
    }

    long long **hashRow = malloc(n * sizeof(long long *));
    for (int i = 0; i < n; i++) {
        hashRow[i] = calloc(n, sizeof(long long));
    }

    long long **hashCol = malloc(n * sizeof(long long *));
    for (int j = 0; j < n; j++) {
        hashCol[j] = calloc(n, sizeof(long long));
    }

    for (int i = 0; i < n; i++) {
        for (int j = 0; j < n; j++) {
            for (int k = 0; k < n; k++) {
                int mod_jk = (j - k + n) % n;
                long long term = (unsigned char)a[i][j] * powP[mod_jk];
                hashRow[i][k] += term;

                int mod_ik = (i - k + n) % n;
                term = (unsigned char)a[i][j] * powP[mod_ik];
                hashCol[j][k] += term;
            }
        }
    }

    int ans = 0;
    for (int A = 0; A < n; A++) {
        for (int B = 0; B < n; B++) {
            int ok = 1;
            for (int i = 0; i < n; i++) {
                int row_iA = (i + A) % n;
                int col_iB = (i + B) % n;
                if (hashRow[row_iA][B] != hashCol[col_iB][A]) {
                    ok = 0;
                    break;
                }
            }
            ans += ok;
        }
    }

    printf("%d\n", ans);

    for (int i = 0; i < n; i++) free(a[i]);
    free(a);
    free(powP);
    for (int i = 0; i < n; i++) free(hashRow[i]);
    free(hashRow);
    for (int j = 0; j < n; j++) free(hashCol[j]);
    free(hashCol);

    return 0;
}
// End of Code
